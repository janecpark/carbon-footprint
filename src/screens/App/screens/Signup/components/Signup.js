import React, { useState, useContext } from 'react';
import UserApi from '../../../UserApi';
import { useHistory } from 'react-router-dom';
import TokenContext from '../../../TokenContext';
import Alert from '../../../../../shared/components/Alert';

/** Sign up form that registers users and sets token in localStorage */

const Signup = () => {
  const { setToken } = useContext(TokenContext);
  const history = useHistory();
  const INITIAL_STATE = {
    username: '',
    password: '',
    email: '',
    errors: [],
  };
  const [userData, setUserData] = useState(INITIAL_STATE);
  const handleChange = (e) => {
    const { name, value } = e.target;
    setUserData((el) => ({
      ...el,
      [name]: value,
    }));
  };

  async function handleSubmit(e) {
    e.preventDefault();
    let data = {
      username: userData.username,
      password: userData.password,
      email: userData.email,
    };
    let res;
    try {
      res = await UserApi.signUp(data);
      setToken(res.token);
    } catch (err) {
      return setUserData((f) => ({ ...f, errors: err }));
    }
    history.push('/');
  }
  return (
    <div className="Signup">
      <h2 className='Signup-header'>Sign up</h2>
      <form className="signUpForm container" onSubmit={handleSubmit}>
        <div className="form-group">
          <label htmlFor="username">Username</label>
          <input
            type="text"
            className="form-control"
            id="username"
            name="username"
            onChange={handleChange}
            value={userData.username}
            required
          />
        </div>
        <div className="form-group">
          <label htmlFor="password">Password</label>
          <input
            type="password"
            className="form-control"
            id="password"
            name="password"
            onChange={handleChange}
            value={userData.password}
            minLength="5"
            required
          />
        </div>
        <div className="form-group">
          <label htmlFor="email">Email</label>
          <input
            type="email"
            className="form-control"
            id="email"
            name="email"
            onChange={handleChange}
            value={userData.email}
            required
          />
        </div>
        <button className="btn btn-primary btn-block">Sign Up</button>
        {userData.errors.length ? (
          <Alert type="danger" messages={userData.errors} />
        ) : null}
      </form>
    </div>
  );
};

export default Signup;
