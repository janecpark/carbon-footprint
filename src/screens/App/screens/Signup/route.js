import { routeConstants } from '../../../../shared/constants';
import Signup from "./";

export default {
    path: routeConstants.SIGNUP.route,
    component: Signup
};