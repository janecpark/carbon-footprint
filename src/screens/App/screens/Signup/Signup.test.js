import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import Signup from './components/Signup';
import {
  UserProvider,
  FormProvider,
  ResultProvider,
  ActionProvider,
  TokenProvider,
} from '../testUtils';

it('renders without crashing', function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Signup />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
});

it('matches snapshot', function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Signup />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders text on page', () => {
  const { getByText } = render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Signup />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
  const element = getByText(/Sign Up/);
  expect(element).toBeInTheDocument();
});
