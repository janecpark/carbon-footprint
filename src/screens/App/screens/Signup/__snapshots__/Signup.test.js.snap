// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Signup"
  >
    <form
      class="signUpForm container"
    >
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          class="form-control"
          id="username"
          name="username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="password"
          minlength="5"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          class="form-control"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <button
        class="btn btn-primary btn-block"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
